### PR TITLE
Update SendMessage and Set/GetWindowsLong

### PR DESCRIPTION
**Pull request by JeremyKuhne**
This changes to a correctly typed WPARAM (unsigned) and LPARAM (signed).

Adds a number of explicit cast operators to try and do the right thing when passing/extracting values through WPARAM, LPARAM, LRESULT.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7586)

### CODE REVIEW

**Review by dreddy-work — APPROVED** [review]
RLGTM.

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 1]
```diff
@@ -3769,13 +3776,13 @@ protected void CenterToScreen()
...
-                IntPtr hWndOwner = IntPtr.Zero;
+                HWND hWndOwner = default;
                 if (TopLevel)
                 {
-                    hWndOwner = User32.GetWindowLong(this, User32.GWL.HWNDPARENT);
+                    hWndOwner = (HWND)PInvoke.GetWindowLong(this, WINDOW_LONG_PTR_INDEX.GWL_HWNDPARENT);
                 }
 
-                desktop = hWndOwner != IntPtr.Zero ? Screen.FromHandle(hWndOwner) : Screen.FromPoint(MousePosition);
+                desktop = !hWndOwner.IsNull ? Screen.FromHandle(hWndOwner) : Screen.FromPoint(MousePosition);
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:3785]: Normally yes. Due to the scope of this change I was trying to minimize the diff.

File: src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs [hunk 2]
```diff
@@ -1349,14 +1349,14 @@ private bool CheckSubclassing()
                 return true;
             }
 
-            IntPtr handle = Handle;
-            IntPtr currentWndproc = User32.GetWindowLong(this, User32.GWL.WNDPROC);
+            HWND handle = HWND;
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs:1352]: It is used below with `WindowAssignHandle(handle)`

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerActionPanel.EditorPropertyLine.cs [hunk 3]
```diff
@@ -565,17 +565,17 @@ public void DoModalLoop()
...
-                    while (hWnd != IntPtr.Zero)
+                    while (!hWnd.IsNull)
                     {
-                        hWnd = User32.GetWindowLong(hWnd, User32.GWL.HWNDPARENT);
-                        if (hWnd == IntPtr.Zero)
+                        hWnd = (HWND)PInvoke.GetWindowLong(hWnd, WINDOW_LONG_PTR_INDEX.GWL_HWNDPARENT);
+                        if (hWnd.IsNull)
                         {
                             return false;
                         }
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerActionPanel.EditorPropertyLine.cs:577]: Not sure why you think this can be removed, it would change the logic here. Can you elaborate?

**Comment by JeremyKuhne** [comment]
x64 release appears to have timed out

File: src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs:1352]: That is correct and that is the only place it is used. `HWD` can be used directly?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerActionPanel.EditorPropertyLine.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerActionPanel.EditorPropertyLine.cs:577]: While loop is checking and bailout for next iteration if it is null. Am i missing something?

File: src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs:1352]: Possibly, but that would have to be thoroughly investigated for side effects. Getting it here caches the value before the other calls, I had to presume that there was a reason (it might change).

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerActionPanel.EditorPropertyLine.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerActionPanel.EditorPropertyLine.cs:577]: It would change the semantics here. `if (hWnd == (HWND)Handle)` could create the handle now when it might not have before. Best to be very careful about modifying logic around `.Handle` access.

Side lesson here is that it is terrible design to have properties with side effects. `.Handle` should never have created the handle, there should have been a `GetOrCreateHandle()` or something similar.

File: src/System.Windows.Forms.Primitives/src/NativeMethods.txt [hunk 4]
```diff
@@ -91,11 +98,12 @@ NMHDR
 PeekMessage
 PostMessage
 RegisterClass
+SendMessage
 SetActiveWindow
 SetFocus
 SetParent
 SetSuspendState
-SetWindowLong
```
Comment (lonitra) [inline src/System.Windows.Forms.Primitives/src/NativeMethods.txt:98]: Why was this removed?

Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/NativeMethods.txt:98]: I can't figure out how to get `SetWindowLongPtr` to map in and additionally I felt it was better to keep these manual and private to avoid accidentally calling the wrong one when running 32/64 bit.

Comment (lonitra) [reply src/System.Windows.Forms.Primitives/src/NativeMethods.txt:98]: Gotcha, that makes sense. I did a quick search and found a seemingly related [discussion](https://github.com/microsoft/win32metadata/issues/384). Maybe there are clues here, but we can investigate more in the future.

**Review by lonitra — APPROVED** [review]
:shipit:
